File: engine/test/content/InstrumentConfigTest.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <gtest/gtest.h>

#include "xjmusic/content/InstrumentConfig.h"
#include "xjmusic/util/StringUtils.h"

using namespace XJ;

TEST(InstrumentConfigTest, SetFromDefaults) {
  const InstrumentConfig subject;

  ASSERT_FALSE(subject.isMultiphonic);
}

TEST(InstrumentConfigTest, DefaultsToString) {
  const InstrumentConfig subject;

  const std::vector<std::string> defaultLines = StringUtils::split(InstrumentConfig::DEFAULT, '\n');
  const std::vector<std::string> subjectLines = StringUtils::split(subject.toString(), '\n');

  ASSERT_EQ(defaultLines, subjectLines);
}

TEST(InstrumentConfigTest, IsOneShot) {
  const InstrumentConfig subject("isOneShot=true");

  ASSERT_TRUE(subject.isOneShot);
}

TEST(InstrumentConfigTest, OneShotObserveLengthOfEvents) {
  const InstrumentConfig subject("oneShotObserveLengthOfEvents=[   bada ,     bIng, b    ooM ]");

  const std::vector<std::string> expected = {"BADA", "BING", "BOOM"};
  ASSERT_EQ(expected, subject.oneShotObserveLengthOfEvents);
}

TEST(InstrumentConfigTest, IsOneShotCutoffEnabled) {
  const InstrumentConfig subject("");

  ASSERT_TRUE(subject.isOneShotCutoffEnabled);
}

TEST(InstrumentConfigTest, ReleaseMillis) {
  const InstrumentConfig subject("");

  ASSERT_EQ(5, subject.releaseMillis);
}

TEST(InstrumentConfigTest, IsTonal) {
  const InstrumentConfig subject("isTonal=true");

  ASSERT_TRUE(subject.isTonal);
}

TEST(InstrumentConfigTest, OneShotObserveLengthOfEventsContains) {
  const InstrumentConfig subject("oneShotObserveLengthOfEvents=[   bada ,     bIng, b    ooM ]");

  ASSERT_TRUE(subject.oneShotObserveLengthOfEventsContains("BADA"));
  ASSERT_TRUE(subject.oneShotObserveLengthOfEventsContains("BING"));
  ASSERT_TRUE(subject.oneShotObserveLengthOfEventsContains("BOOM"));
  ASSERT_FALSE(subject.oneShotObserveLengthOfEventsContains("BEEP"));
}

File: workstation/src/test/java/io/xj/model/InstrumentConfigTest.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

package io.xj.model;

import io.xj.model.pojos.Instrument;
import io.xj.model.util.ValueException;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 Artist saves Instrument config, validate & combine with defaults. https://github.com/xjmusic/xjmusic/issues/206
 */
public class InstrumentConfigTest {

  @Test
  public void setFromInstrument() throws ValueException {
    var instrument = new Instrument();
    instrument.setConfig("isMultiphonic = true");

    var subject = new InstrumentConfig(instrument);

    assertTrue(subject.isMultiphonic());
  }

  @Test
  public void setFromInstrument_isAudioSelectionPersistent_false() throws ValueException {
    var instrument = new Instrument();
    instrument.setConfig("isAudioSelectionPersistent = false");

    var subject = new InstrumentConfig(instrument);

    assertFalse(subject.isAudioSelectionPersistent());
  }

  @Test
  public void setFromDefaults() throws ValueException {
    var subject = new InstrumentConfig(InstrumentConfig.DEFAULT);

    assertFalse(subject.isMultiphonic());
  }

  @Test
  public void defaultsToString() throws ValueException {
    var subject = new InstrumentConfig(InstrumentConfig.DEFAULT);

    assertArrayEquals(InstrumentConfig.DEFAULT.split("\n"), subject.toString().split(System.lineSeparator()));
  }

  /**
   Instruments have a config to play back as one-shots (ignore note-event lengths) https://github.com/xjmusic/xjmusic/issues/224
   */
  @Test
  public void isOneShot() throws ValueException {
    var subject = new InstrumentConfig("isOneShot=true");

    assertTrue(subject.isOneShot());
  }

  /**
   Instruments have a config to play back as one-shots (ignore note-event lengths) https://github.com/xjmusic/xjmusic/issues/224
   */
  @Test
  public void oneShotObserveLengthOfEvents() throws ValueException {
    var subject = new InstrumentConfig("oneShotObserveLengthOfEvents=[   bada ,     bIng, b    ooM ]");

    assertEquals(List.of("BADA", "BING", "BOOM"), subject.getOneShotObserveLengthOfEvents());
  }

  /**
   InstrumentConfig to disable note cutoffs for one-shot instruments https://github.com/xjmusic/xjmusic/issues/225
   */
  @Test
  public void isOneShotCutoffEnabled() throws ValueException {
    var subject = new InstrumentConfig("");

    assertTrue(subject.isOneShotCutoffEnabled());
  }

  /**
   One-shot fadeout mode https://github.com/xjmusic/xjmusic/issues/226
   */
  @Test
  public void releaseMillis() throws ValueException {
    var subject = new InstrumentConfig("");

    assertEquals((Integer) 5, subject.getReleaseMillis());
  }

}
